### PR TITLE
calib3d: check for minimum corresponding points to calculate homography

### LINKED ISSUE #18254: Proper extrapolation with findHomography
- OpenCV => 4.4.0.42
- Operating System / Platform => Ubuntu 18.04

There is this simple case I plotted, where the source points of the homography is a simple square side==100 and the destination points is a square with side==10 
(0,0) are the same in both planes
Inferring the homography matrix with 'findHomography' works fine here
So this is the plot with some other points inferred from 'perspectiveTransform' function:
[
Captura de tela de 2020-09-02 11-56-30
](url)
This is straightforward correct, as the red points is kinda notable solutions

Point is, this system shouldn't be easily extrapolated given 3 points, or am I assuming constraints?
Removing the source point in [100,100] gives this answers:
Captura de tela de 2020-09-02 11-57-23
How can it be so different from the obvious solution from the first case?
It is well documented that if the matchs aren't rigid, its estimation is poor, is this case not rigid?
It would be great if this was properly extrapolated for a rhombus shape

Passing a non-collinear 4th point anywhere gives the clean solution as well
Captura de tela de 2020-09-02 11-58-45

code(python):

```
import numpy as np
import cv2 as cv
import matplotlib.pyplot as plt

anchors = np.array([[0,0],[100,0],[0,100],[100,100]] ,dtype=np.float32)
values = np.array([[0,0],[10,0],[0,10],[10,10]] ,dtype=np.float32)
grid_points = np.array( [[[100,50] , [50,50] , [50,0] , [25,25]]] ,dtype=np.float32)
A,mask = cv.findHomography( anchors , values )
grid_values = cv.perspectiveTransform( grid_points , A )

#PyPlot'ing
plt.figure(figsize=(10,10))
plt.plot( anchors[:,0] , anchors[:,1] , '.' , ms=20 )
for i in range(0,len(anchors)):
    plt.text(anchors[i,0]-1,anchors[i,1]+5, str(values[i]) )

plt.plot( grid_points[0,:,0]   , grid_points[0,:,1]   , '.' , ms=5 )
for i in range(0,len(grid_points[0])):
    plt.text( grid_points[0,i,0]-1 , grid_points[0,i,1]-5 , str('{:.2f},{:.2f}'.format(grid_values[0,i,0],grid_values[0,i,1])) , color='red')

plt.show()

```

### PR DESCRIPTION
**Pull request by aitikgupta**
Fixes #18254 
- Added a check for minimum points to calculate Homography

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
allow_multiple_commits=1
```

### CODE REVIEW

File: modules/calib3d/src/fundam.cpp [hunk 1]
```diff
@@ -377,7 +377,8 @@ cv::Mat cv::findHomography( InputArray _points1, InputArray _points2,
         p.reshape(2, npoints).convertTo(m, CV_32F);
     }
 
-    CV_Assert( src.checkVector(2) == dst.checkVector(2) );
+    // Need atleast 4 point correspondences
+    CV_Assert( src.checkVector(2) == dst.checkVector(2) && src.checkVector(2) > 3 );
```
Comment (alalek) [inline modules/calib3d/src/fundam.cpp:381]: 1. Cache expression result: `src.checkVector(2)` is used twice - no need to evaluate it twice.
2. Avoid expanding of assertion check. Such error messages are not user friendly. Use one check per assertion or better to use `CV_Check*()` macros.
3. Please add simple test for this case.

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

Please take a look on comment below.

File: modules/calib3d/src/fundam.cpp (hunk 1 above)
Comment (aitikgupta) [reply modules/calib3d/src/fundam.cpp]: I've updated the commit with a better check here

**Comment by aitikgupta** [comment]
@alalek Can you review again? Thanks

File: modules/calib3d/src/fundam.cpp [hunk 2]
```diff
@@ -374,6 +374,9 @@ cv::Mat cv::findHomography( InputArray _points1, InputArray _points2,
                 return Mat();
             convertPointsFromHomogeneous(p, p);
         }
+        // Need atleast 4 point correspondences to calculate Homography
```
Comment (catree) [inline modules/calib3d/src/fundam.cpp:377]:
```suggestion
        // Need at least 4 point correspondences to calculate Homography
```

File: modules/calib3d/src/fundam.cpp [hunk 3]
```diff
@@ -374,6 +374,9 @@ cv::Mat cv::findHomography( InputArray _points1, InputArray _points2,
...
+        if( npoints < 4 )
+            CV_Error(Error::StsVecLengthErr , "The input arrays should have atleast 4 corresponding point sets to calculate Homography");
```
Comment (catree) [inline modules/calib3d/src/fundam.cpp:379]:
```suggestion
            CV_Error(Error::StsVecLengthErr , "The input arrays should have at least 4 corresponding point sets to calculate Homography");
```

File: modules/calib3d/test/test_homography.cpp [hunk 4]
```diff
@@ -711,4 +712,29 @@ TEST(Calib3d_Homography, fromImages)
     ASSERT_GE(ninliers1, 80);
 }
 
+TEST(Calib3d_Homography, minPoints)
```
Comment (catree) [inline modules/calib3d/test/test_homography.cpp:715]: I guess a supplementary test with input points < 4 and check that an exception is correctly thrown should be added?

**Comment by aitikgupta** [comment]
@catree made the changes :+1:

**Comment by aitikgupta** [comment]
[This 1 build](https://pullrequest.opencv.org/buildbot/builders/precommit_opencl/builds/26039) timed out after 40 minutes. Not sure how to proceed?

**Review by alalek — APPROVED** [review]
Thank you :+1:

**Pull request merged**